### ISSUE TITLE
Equal val_mean_pos and val_mean_neg during training

### ISSUE THREAD

**Issue by DanilZittser** [tensorflow/similarity#231]
Hi everyone!

After upgrade tfsim to 0.15.2 and start using `EfficientNetSim` i have same mean_pos and mean_neg on validation set:
```
Epoch 2/128
256/256 [==============================] - 57s 225ms/step - loss: 1.5161 - mean_pos: 0.0641 - mean_neg: 0.1761 - val_loss: 1.0291 - val_mean_pos: 0.0828 - val_mean_neg: 0.0828
Epoch 3/128
256/256 [==============================] - 57s 222ms/step - loss: 1.3810 - mean_pos: 0.0529 - mean_neg: 0.1934 - val_loss: 0.8178 - val_mean_pos: 0.0771 - val_mean_neg: 0.0771
Epoch 4/128
256/256 [==============================] - 56s 218ms/step - loss: 1.2050 - mean_pos: 0.0454 - mean_neg: 0.2100 - val_loss: 0.7022 - val_mean_pos: 0.0855 - val_mean_neg: 0.0855
Epoch 5/128
256/256 [==============================] - 56s 217ms/step - loss: 1.0587 - mean_pos: 0.0417 - mean_neg: 0.2269 - val_loss: 0.6784 - val_mean_pos: 0.0712 - val_mean_neg: 0.0712
Epoch 6/128
256/256 [==============================] - 56s 217ms/step - loss: 0.9288 - mean_pos: 0.0385 - mean_neg: 0.2400 - val_loss: 0.5951 - val_mean_pos: 0.0586 - val_mean_neg: 0.0586
Epoch 7/128
256/256 [==============================] - 56s 217ms/step - loss: 0.8379 - mean_pos: 0.0359 - mean_neg: 0.2513 - val_loss: 0.4841 - val_mean_pos: 0.0930 - val_mean_neg: 0.0930
Epoch 8/128
256/256 [==============================] - 56s 217ms/step - loss: 0.6939 - mean_pos: 0.0347 - mean_neg: 0.2650 - val_loss: 0.5244 - val_mean_pos: 0.0813 - val_mean_neg: 0.0813
Epoch 9/128
256/256 [==============================] - 56s 219ms/step - loss: 0.6005 - mean_pos: 0.0333 - mean_neg: 0.2800 - val_loss: 0.4762 - val_mean_pos: 0.0842 - val_mean_neg: 0.0842
...
```

Could you please suggest a possible reason for this behavior?

Use this model:
```python
model = tfsim.architectures.EfficientNetSim(
    input_shape=(224 224, 3),
    embedding_size=128,
    weights='imagenet',
    trainable='partial',
    l2_norm=True,
    include_top=True,
    pooling='max',    # Can change to use `gem` -> GeneralizedMeanPooling2D
    gem_p=3.0,        # Increase the contrast between activations in the feature map.
)

model.compile(
    optimizer=tf.keras.optimizers.Adam(0.0001),
    loss=tfsim.losses.MultiSimilarityLoss(distance='cosine'),
    metrics=[
        tfsim.training_metrics.avg_pos(distance='cosine'),
        tfsim.training_metrics.avg_neg(distance='cosine'),
    ],
)

history = model.fit(sampler, epochs=128, validation_data=(x_test, y_test))
```

Test set contain 60 classes and 24 samples per class.

**Comment by owenvallis** [comment]
Hi @DanilZittser, thanks for raising this issue. I'll try and take a look into it this week.

**Comment by DanilZittser** [comment]
@owenvallis, looking forward to it!

TodayI started training the model in the Yandex.Cloud and get the same problem:

```
...
Epoch 12/64
256/256 [==============================] - 145s 566ms/step - loss: 0.4077 - mean_pos: 0.0194 - mean_neg: 0.2479 - val_loss: 0.4174 - val_mean_pos: 0.3327 - val_mean_neg: 0.3327
Epoch 13/64
256/256 [==============================] - 145s 567ms/step - loss: 0.3559 - mean_pos: 0.0174 - mean_neg: 0.2489 - val_loss: 0.4322 - val_mean_pos: 0.3646 - val_mean_neg: 0.3646
Epoch 14/64
256/256 [==============================] - 145s 566ms/step - loss: 0.3199 - mean_pos: 0.0154 - mean_neg: 0.2508 - val_loss: 0.4136 - val_mean_pos: 0.3415 - val_mean_neg: 0.3415
Epoch 15/64
256/256 [==============================] - 145s 567ms/step - loss: 0.2805 - mean_pos: 0.0155 - mean_neg: 0.2600 - val_loss: 0.4218 - val_mean_pos: 0.3268 - val_mean_neg: 0.3269
Epoch 16/64
256/256 [==============================] - 145s 566ms/step - loss: 0.2463 - mean_pos: 0.0143 - mean_neg: 0.2526 - val_loss: 0.4155 - val_mean_pos: 0.3323 - val_mean_neg: 0.3323
...
```
Today i used this versions of packages:
```
tensorflow==2.6.0
tensorflow-datasets==4.5.2
tensorflow-estimator==2.7.0
tensorflow-metadata==1.5.0
tensorflow-similarity==0.15.2
```

In first my post i used this version:
```
tensorflow==2.4.4
tensorflow-datasets==4.4.0
tensorflow-estimator==2.4.0
tensorflow-io-gcs-filesystem==0.23.1
tensorflow-metadata==1.5.0
tensorflow-similarity==0.15.2
```

**Comment by owenvallis** [comment]
Hi @DanilZittser,

Since this is just happening on the validation set, it might be that the model isn't generalizing and is unable to separate the validation examples. Can you try adding the min_ and max_ training metrics as well?

**Comment by DanilZittser** [comment]
Hello, @owenvallis!

The problem was that the validation data was sorted by labels, so the batch rarely consisted of different labels.

How correct is it to use `MultiShotMemorySampler` for validation data?

**Comment by owenvallis** [comment]
Hmm, that's odd though, the sampler randomly samples elements from each class to ensure we get at least a pair of examples per class per batch. Can you share the size of the contiguous blocks of class labels in y_validate and the values you had for:

* env.classes_per_batch
* env.examples_per_class_per_batch
* env.steps_per_epoch

I'd like to check the batch outputs of the MultiShotMemorySampler.

**Comment by DanilZittser** [comment]
```python
env.classes_per_batch = 128
env.examples_per_class_per_batch = 4
env.steps_per_epoch = 1024
```

I using `Sampler` only for train data, validation data was simple tuples (`(x[test_mask, ...], y[test_mask, ...])`) sorted by labels.

**Comment by owenvallis** [comment]
Ah, I see! Yeah, this would totally cause the issue you saw.

Glad to see you isolated the issue and let us know if anything else comes up.

Cheers,
